core/arch/emit_utils_shared.c

```diff
@@ -5534,6 +5536,56 @@ emit_special_ibl_xfer(dcontext_t *dcontext, byte *pc, generated_code_t *code, ui
         in = instrlist_first(custom_ilist);
     }
 
+#ifdef UNIX
+    /* i#4670: Jump to the unlinked IBL target if there are pending signals. This is
+     * required to bound delivery time for signals received while executing fragments
+     * that use the special ibl xfer trampoline, which uses a different (un)linking
+     * mechanism.
+     */
+    insert_shared_get_dcontext(dcontext, &ilist, NULL, true);
+#    ifdef X86
+    APP(&ilist,
+        XINST_CREATE_cmp(dcontext,
```

Comment:
This clobbers the flags but I don't see them saved?  I assume you're not assuming -unsafe_ignore_eflags_ibl (though this is not always an app ib).  Use load+jecxz?  Requires swapping regs.

Reply:
Do you mean that we should load `dcontext` into `xcx` (instead of `xdi`) to avoid load/store for a second scratch reg? We can then reuse `ecx` to load pending signals count, and then restore only `xcx`.

Reply:
Or is it better to just reuse `insert_shared_get_dcontext` as-is (it has some extra logic for `-protect_mask`) to load `dcontext` into `xdi` and use `xcx` as scratch?

Reply:
Or load into xdi and xchg with xcx, and xchg back on both paths I guess.  xchg reg-to-reg is fast I believe; only w/ memory is it to be avoided.

Reply:
Done.